fhir-resource-tooling/src/main/java/com/ibm/cohort/tooling/fhir/ValueSetImporter.java

```diff
@@ -0,0 +1,161 @@
+package com.ibm.cohort.tooling.fhir;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.compress.utils.Lists;
+import org.apache.poi.ss.usermodel.CellType;
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.xssf.usermodel.XSSFSheet;
+import org.apache.poi.xssf.usermodel.XSSFWorkbook;
+import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.MetadataResource;
+import org.hl7.fhir.r4.model.ValueSet;
+
+import com.beust.jcommander.JCommander;
+import com.beust.jcommander.Parameter;
+import com.beust.jcommander.internal.Console;
+import com.beust.jcommander.internal.DefaultConsole;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.ibm.cohort.fhir.client.config.FhirClientBuilderFactory;
+import com.ibm.cohort.fhir.client.config.FhirServerConfig;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.parser.IParser;
+import ca.uhn.fhir.rest.api.MethodOutcome;
+import ca.uhn.fhir.rest.client.api.IGenericClient;
+
+public class ValueSetImporter {
+
+	public static final class ValueSetImporterArguments {
+		@Parameter(names = {"-m",
+				"--measure-server"}, description = "Path to JSON configuration data for the FHIR server connection that will be used to retrieve measure and library resources.", required = true)
+		File measureServerConfigFile;
+
+		@Parameter(names = {"-h", "--help"}, description = "Show this help", help = true)
+		boolean isDisplayHelp;
+
+		@Parameter(description = "The list of value set spreadsheets to import", required = true)
+		List<String> spreadsheets;
+	}
+
+	private static class ValueSetArtifact {
+
+		MetadataResource resource;
+		String url;
+		String name;
+		String id;
+	}
+
+	private IGenericClient client;
+	private IParser parser;
+
+	private ValueSetImporter(IGenericClient client) {
+		this.client = client;
+		this.parser = client.getFhirContext().newJsonParser();
+	}
+
+	private void importArtifacts(List<ValueSetArtifact> valueSetArtifacts) {
+
+		for (ValueSetArtifact valueSetArtifact : valueSetArtifacts) {
+			Bundle bundle = client.search().forResource(ValueSet.class).where(ValueSet.URL.matches().value(valueSetArtifact.url))
+					.returnBundle(Bundle.class).execute();
+			if (bundle.getEntry().size() > 0) {
+				valueSetArtifact.id = bundle.getEntryFirstRep().getResource().getIdElement().getIdPart();
+			} else {
+				MethodOutcome outcome = client.create().resource(parser.encodeResourceToString(valueSetArtifact.resource)).execute();
+				if (outcome.getCreated()) {
+					valueSetArtifact.id = outcome.getId().getIdPart();
+				}
+			}
+		}
+	}
+
+	private ValueSetArtifact createArtifact(InputStream is) throws IOException {
+		XSSFWorkbook wb = new XSSFWorkbook(is);
+		XSSFSheet mainSheet = wb.getSheetAt(0);
+		ValueSet valueSet = new ValueSet();
```

Comment:
Would it make sense to also verify the "tab label" for each sheet - e.g. in case user inserted another worksheet in slot 0 or 1? Or just assume we will encounter other errors if the sheet "n" doesn't contain the information we're looking for?

Reply:
If we think the sheet names are more stable than their positions, I'm happy to swap to retrieving by that instead